pkg/controller/plan/adapter/ocp/builder.go

```diff
@@ -266,33 +268,47 @@ func (r *Builder) mapDisks(sourceVm *cnv.VirtualMachine, targetVmSpec *cnv.Virtu
 		}
 	}
 
+	diskMap := createDiskMap(sourceVm, pvcMap, vmRef)
+	configMaps, secrets := r.createEnvMaps(sourceVm, vmRef)
+
+	// Clear original disks and volumes, will be required for other mapped devices later
+	targetVmSpec.Template.Spec.Domain.Devices.Disks = []cnv.Disk{}
+	targetVmSpec.Template.Spec.Volumes = []cnv.Volume{}
+
+	r.mapPVCsToTarget(targetVmSpec, persistentVolumeClaims, diskMap)
+	r.mapConfigMapsToTarget(targetVmSpec, configMaps, diskMap)
+	r.mapSecretsToTarget(targetVmSpec, secrets, diskMap)
+}
+
+func createDiskMap(sourceVm *cnv.VirtualMachine, pvcMap map[string]*core.PersistentVolumeClaim, vmRef ref.Ref) map[string]*cnv.Disk {
 	diskMap := make(map[string]*cnv.Disk)
-	for i := range sourceVm.Spec.Template.Spec.Domain.Devices.Disks {
-		disk := &sourceVm.Spec.Template.Spec.Domain.Devices.Disks[i]
-		// Find matching volume
-		// TODO: make this better
+
+	for _, disk := range sourceVm.Spec.Template.Spec.Domain.Devices.Disks {
+		currentDisk := disk
 		for _, vol := range sourceVm.Spec.Template.Spec.Volumes {
-			if vol.PersistentVolumeClaim != nil {
-				if vol.Name == disk.Name {
-					key := pvcSourceName(vmRef.Namespace, vol.PersistentVolumeClaim.ClaimName)
-					if _, ok := pvcMap[key]; ok {
-						diskMap[key] = disk
-						continue
-					}
-				}
-			} else if vol.DataVolume != nil {
-				if vol.Name == disk.Name {
-					key := pvcSourceName(vmRef.Namespace, vol.DataVolume.Name)
-					if _, ok := pvcMap[key]; ok {
-						diskMap[key] = disk
-						continue
-					}
-				}
+			var key string
+			switch {
+			case vol.PersistentVolumeClaim != nil:
+				key = pvcSourceName(vmRef.Namespace, vol.PersistentVolumeClaim.ClaimName)
+			case vol.DataVolume != nil:
+				key = pvcSourceName(vmRef.Namespace, vol.DataVolume.Name)
+			case vol.ConfigMap != nil:
+				key = vol.ConfigMap.Name
+			case vol.Secret != nil:
+				key = vol.Secret.SecretName
+			}
+			if vol.Name == disk.Name {
```

Comment:
that's a bit inefficient to go through the switch-case block above and only then check this, no? how about moving it to the beginning of the for-loop?